sphinx/ext/autodoc/__init__.py

```diff
@@ -693,17 +693,26 @@ def document_members(self, all_members: bool = False) -> None:
         member_order = self.options.member_order or \
             self.env.config.autodoc_member_order
         if member_order == 'groupwise':
-            # sort by group; relies on stable sort to keep items in the
-            # same group sorted alphabetically
-            memberdocumenters.sort(key=lambda e: e[0].member_order)
-        elif member_order == 'bysource' and self.analyzer:
-            # sort by source order, by virtue of the module analyzer
-            tagorder = self.analyzer.tagorder
-
-            def keyfunc(entry: Tuple[Documenter, bool]) -> int:
-                fullname = entry[0].name.split('::')[1]
-                return tagorder.get(fullname, len(tagorder))
-            memberdocumenters.sort(key=keyfunc)
+            # sort by group; alphabetically within groups
+            memberdocumenters.sort(key=lambda e: (e[0].member_order, e[0].name))
+        elif member_order == 'bysource':
+            if self.analyzer:
+                # sort by source order, by virtue of the module analyzer
+                tagorder = self.analyzer.tagorder
+
+                def keyfunc(entry: Tuple[Documenter, bool]) -> int:
+                    fullname = entry[0].name.split('::')[1]
+                    return tagorder.get(fullname, len(tagorder))
+                memberdocumenters.sort(key=keyfunc)
+            else:
+                # Assume that member discovery order matches source order.
+                # This is a reasonable assumption in Python 3.6 and up, where
+                # module.__dict__ is insertion-ordered.
+                pass
```

Comment:
We need to consider the order of the members not included in `__dict__`.

* enum members (they are not in `__dict__`)
* `__slots__`
* `__annotations__` (type annotated variables, but not initialized yet)
* instance variables
* members defined at ancestor classes
* `module.__all__` (which is strong `__dict__` or `__all__`?)

Note: Now we don't use `__dict__` for discovery. So we have to order them manually if we adopt `__dict__`.

Reply:
Bear in mind this is just a backup option for when no analyzer is available. You're right that this is not a perfect ordering, but sorting alphabetically when "by source" is requested is far worse than a rough approximation to by source.

---

> enum members (they are not in __dict__)

I disagree:
```
>>> C = enum.Enum("Color", "red green blue")
>>> C.__dict__\
mappingproxy({'_generate_next_value_': <function enum.Enum._generate_next_value_(name, start, count, last_values)>,
              '__doc__': 'An enumeration.',
              '__module__': '__main__',
              '_member_names_': ['red', 'green', 'blue'],
              '_member_map_': {'red': <Color.red: 1>,
               'green': <Color.green: 2>,
               'blue': <Color.blue: 3>},
              '_member_type_': object,
              '_value2member_map_': {1: <Color.red: 1>,
               2: <Color.green: 2>,
               3: <Color.blue: 3>},
              'red': <Color.red: 1>,
              'green': <Color.green: 2>,
              'blue': <Color.blue: 3>,
              '__new__': <function enum.Enum.__new__(cls, value)>})
```

> `__slots__`

Also not true:
```
>>> class Foo:
...     __slots__ = ('a', 'b')

>>> Foo.__dict__
mappingproxy({'__module__': '__main__',
              '__slots__': ('a', 'b'),
              'a': <member 'a' of 'Foo' objects>,
              'b': <member 'b' of 'Foo' objects>,
              '__doc__': None})
```

> `__annotations__` (type annotated variables, but not initialized yet)

True, but likely irrelevant for C extension modules

> instance variables

True

> members defined at ancestor classes

True

> `module.__all__`

Likely irrelevant for C extension modules

Reply:
>enum members (they are not in dict)

Sorry, this is my wrong. I mistook `dir()` for `__dict__`. The result of `dir()` for enum classes does not contain methods. But it is not related to sorting. Please forget.
```
Python 3.8.3 (default, May 15 2020, 11:21:44)
[Clang 11.0.3 (clang-1103.0.32.59)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import enum
>>> class MyEnum(enum.Enum):
...     foo = 1
...     bar = 2
...     def hello(self): pass
...
>>> dir(MyEnum)
['__class__', '__doc__', '__members__', '__module__', 'bar', 'foo']
```

> __slots__

This is also my wrong.

Reply:
Note: We have some issues for sorting members: https://github.com/sphinx-doc/sphinx/issues/5270, https://github.com/sphinx-doc/sphinx/issues/3673, https://github.com/sphinx-doc/sphinx/pull/7633, https://github.com/sphinx-doc/sphinx/pull/7177 and so on.
So I'm thinking about the whole of sorting system for autodoc.

But they are not a blocker of this improvement. So no reason to merge this. I'll review this again later (maybe tonight).

Reply:
Thanks for those links, I agree that this is not an end-all solution, but I think removing the initial alphabetic sort is probably a good first step.

I've rebased this PR after your `alphabetical` spelling fix.